diff --git a/engine/platform/src/platform_window_glfw3.cpp b/engine/platform/src/platform_window_glfw3.cpp
index d6111293ae..fc8c80ef9c 100644
--- a/engine/platform/src/platform_window_glfw3.cpp
+++ b/engine/platform/src/platform_window_glfw3.cpp
@@ -153,6 +153,10 @@ namespace dmPlatform
         if (glfwInit() == GL_FALSE)
         {
             glfwInitHint(GLFW_PLATFORM, GLFW_PLATFORM_WAYLAND);
+            // TEMPORARY HACK
+            glfwInitHint(GLFW_WAYLAND_LIBDECOR, GLFW_WAYLAND_DISABLE_LIBDECOR);
+            dmLogInfo("Trying to initialize glfw with WAYLAND_LIBDECOR disabled");
+            //
             if (glfwInit() == GL_FALSE)
             {
                 dmLogError("Could not initialize glfw.");
diff --git a/scripts/build.py b/scripts/build.py
index 3472145631..6b9f1f2387 100755
--- a/scripts/build.py
+++ b/scripts/build.py
@@ -2340,8 +2340,17 @@ class Configuration(object):
 
 
     def upload_to_archive(self, src_file, dst_path):
-        url = join(self.get_archive_path(), dst_path).replace("\\", "/")
-        self.upload_to_s3(src_file, url)
+        # url = join(self.get_archive_path(), dst_path).replace("\\", "/")
+        # self.upload_to_s3(src_file, url)
+        print("<!> upload_to_archive: src_file %s, dst_path %s" % (src_file, dst_path))
+        dst = os.path.join(self.defold_root, self.dynamo_home, "archive")
+        print("                       dst %s" % dst)
+        self._mkdirs(os.path.join(dst, os.path.dirname(dst_path)))
+        print("                       mkdirs %s" % os.path.join(dst, os.path.dirname(dst_path)))
+        if os.path.isdir(src_file):
+            self._copy_tree(src_file, os.path.join(dst, dst_path))
+        else:
+            self._copy(src_file, os.path.join(dst, dst_path))
 
 
     def download_from_s3(self, path, url):
