diff --git a/build_tools/sdk.py b/build_tools/sdk.py
index c73b4b5bcc..8af74495e6 100644
--- a/build_tools/sdk.py
+++ b/build_tools/sdk.py
@@ -71,8 +71,8 @@ ANDROID_64_NDK_API_VERSION='21' # Android 5.0
 # Win32
 
 # The version we have prepackaged
-VERSION_WINDOWS_SDK_10="10.0.20348.0"
-VERSION_WINDOWS_MSVC_2022="14.37.32822"
+VERSION_WINDOWS_SDK_10="10.0.26100.0"
+VERSION_WINDOWS_MSVC_2022="14.42.34433"
 PACKAGES_WIN32_TOOLCHAIN=f"Microsoft-Visual-Studio-2022-{VERSION_WINDOWS_MSVC_2022}"
 PACKAGES_WIN32_SDK_10=f"WindowsKits-{VERSION_WINDOWS_SDK_10}"
 
diff --git a/ci/ci.py b/ci/ci.py
index 98edb0c7bd..bd4e13a95f 100644
--- a/ci/ci.py
+++ b/ci/ci.py
@@ -243,8 +243,8 @@ def build_engine(platform, channel, with_valgrind = False, with_asan = False, wi
     if with_vanilla_lua:
         waf_opts.append('--use-vanilla-lua')
 
-    if platform == 'x86_64-linux':
-        args.append('build_sdk_headers') # gather headers after a successful build
+    # if platform == 'x86_64-linux':
+    #     args.append('build_sdk_headers') # gather headers after a successful build
 
     cmd = ' '.join(args + opts)
 
@@ -374,7 +374,7 @@ def install_ext(platform = None):
     call("python scripts/build.py install_ext %s" % ' '.join(opts))
 
 def build_bob(channel, branch = None):
-    args = "python scripts/build.py install_sdk install_ext sync_archive build_bob archive_bob".split()
+    args = "python scripts/build.py install_sdk install_ext build_bob archive_bob --skip-tests".split()
     opts = []
     opts.append("--channel=%s" % channel)
 
diff --git a/engine/sound/src/devices/device_js.cpp b/engine/sound/src/devices/device_js.cpp
index 22dbc6e5cc..faa19bbf1f 100644
--- a/engine/sound/src/devices/device_js.cpp
+++ b/engine/sound/src/devices/device_js.cpp
@@ -87,6 +87,7 @@ namespace dmDeviceJS
         assert(info);
         JSDevice *dev = (JSDevice*) device;
         info->m_MixRate = dmGetDeviceSampleRate(dev->devId);
+        info->m_FrameCount = info->m_MixRate * 0.022;
     }
 
     void DeviceJSStart(dmSound::HDevice device)
diff --git a/engine/sound/src/devices/device_wasapi.cpp b/engine/sound/src/devices/device_wasapi.cpp
index 36d527b2af..c3ff55a044 100644
--- a/engine/sound/src/devices/device_wasapi.cpp
+++ b/engine/sound/src/devices/device_wasapi.cpp
@@ -359,6 +359,7 @@ namespace dmDeviceWasapi
 
         info->m_MixRate = device->m_MixFormat->nSamplesPerSec;
         info->m_FrameCount = device->m_FrameCount;
+        dmLogInfo("<!> DeviceWasapi: info->m_MixRate: %d, info->m_FrameCount: %d", info->m_MixRate, info->m_FrameCount);
     }
 
     static void DeviceWasapiStart(dmSound::HDevice _device)
diff --git a/engine/sound/src/js/library_sound.js b/engine/sound/src/js/library_sound.js
index fe7734fa31..37bc8b0e6f 100644
--- a/engine/sound/src/js/library_sound.js
+++ b/engine/sound/src/js/library_sound.js
@@ -34,6 +34,7 @@ var LibrarySoundDevice =
                     shared.audioCtx = new audioCtxCtor();
                 }
             }
+            console.log("<!> sampleRate", shared.audioCtx.sampleRate);
             // Construct web audio device.
             device = {
                 sampleRate: shared.audioCtx.sampleRate,
@@ -55,6 +56,7 @@ var LibrarySoundDevice =
                     return 0;
                 },
                 _queue: function(samples, sample_count) {
+                    console.log("<!> queue", samples, sample_count);
                     var len = sample_count / this.sampleRate;
                     // use real buffer length next time.
                     this.bufferDuration = len;
diff --git a/scripts/build.py b/scripts/build.py
index 81004793eb..42cd4dc966 100755
--- a/scripts/build.py
+++ b/scripts/build.py
@@ -1368,9 +1368,12 @@ class Configuration(object):
         test_dir = join(self.defold_root, 'com.dynamo.cr/com.dynamo.cr.bob.test')
 
         sha1 = self._git_sha1()
+        print("<!> path is %s" % os.path.join(self.dynamo_home, 'archive', sha1))
         if os.path.exists(os.path.join(self.dynamo_home, 'archive', sha1)):
+            print("<!> copy.sh")
             run.env_shell_command(self._form_env(), "./scripts/copy.sh", cwd=bob_dir)
         else:
+            print("<!> copy_local_bob_artefacts")
             self.copy_local_bob_artefacts()
 
         env = self._form_env()
@@ -2136,8 +2139,17 @@ class Configuration(object):
 
 
     def upload_to_archive(self, src_file, dst_path):
-        url = join(self.get_archive_path(), dst_path).replace("\\", "/")
-        self.upload_to_s3(src_file, url)
+        # url = join(self.get_archive_path(), dst_path).replace("\\", "/")
+        # self.upload_to_s3(src_file, url)
+        print("<!> upload_to_archive: src_file %s, dst_path %s" % (src_file, dst_path))
+        dst = os.path.join(self.defold_root, self.dynamo_home, "archive")
+        print("                       dst %s" % dst)
+        self._mkdirs(os.path.join(dst, os.path.dirname(dst_path)))
+        print("                       mkdirs %s" % os.path.join(dst, os.path.dirname(dst_path)))
+        if os.path.isdir(src_file):
+            self._copy_tree(src_file, os.path.join(dst, dst_path))
+        else:
+            self._copy(src_file, os.path.join(dst, dst_path))
 
 
     def download_from_s3(self, path, url):
