diff --git a/README_ARTSIOM.md b/README_ARTSIOM.md
new file mode 100644
index 0000000000..51b2ac3469
--- /dev/null
+++ b/README_ARTSIOM.md
@@ -0,0 +1,8 @@
+
+./scripts/build.py shell
+./scripts/build.py install_ext
+./scripts/build.py build_engine --platform=arm64-linux --skip-docs --skip-builtins
+./scripts/build.py build_engine --platform=arm64-linux --skip-docs --skip-tests --skip-builtins -- --skip-build-tests
+
+./scripts/build.py build_engine --platform=x86_64-linux --skip-docs --skip-builtins
+./scripts/build.py build_engine --platform=x86_64-linux --skip-docs --skip-tests --skip-builtins -- --skip-build-tests
diff --git a/build_tools/sdk.py b/build_tools/sdk.py
index fec511ff99..c9664731e5 100644
--- a/build_tools/sdk.py
+++ b/build_tools/sdk.py
@@ -71,8 +71,8 @@ ANDROID_64_NDK_API_VERSION='21' # Android 5.0
 # Win32
 
 # The version we have prepackaged
-VERSION_WINDOWS_SDK_10="10.0.20348.0"
-VERSION_WINDOWS_MSVC_2022="14.37.32822"
+VERSION_WINDOWS_SDK_10="10.0.26100.0"
+VERSION_WINDOWS_MSVC_2022="14.42.34433"
 PACKAGES_WIN32_TOOLCHAIN=f"Microsoft-Visual-Studio-2022-{VERSION_WINDOWS_MSVC_2022}"
 PACKAGES_WIN32_SDK_10=f"WindowsKits-{VERSION_WINDOWS_SDK_10}"
 
diff --git a/build_tools/waf_dynamo.py b/build_tools/waf_dynamo.py
index 4bd9ca09f7..4ee6a01fcc 100644
--- a/build_tools/waf_dynamo.py
+++ b/build_tools/waf_dynamo.py
@@ -75,6 +75,10 @@ def platform_supports_feature(platform, feature, data):
         return platform not in ['js-web', 'wasm-web', 'x86_64-ios']
     if feature == 'opengl_compute':
         return platform not in ['js-web', 'wasm-web', 'x86_64-ios', 'arm64-ios', 'arm64-macos', 'x86_64-macos']
+    if feature == 'opengl':
+        return platform not in ['arm64-linux']
+    if feature == 'opengles':
+        return platform in ['arm64-linux']
     if feature == 'webgpu':
         return platform in ['js-web', 'wasm-web']
     return waf_dynamo_vendor.supports_feature(platform, feature, data)
@@ -383,6 +387,7 @@ def default_flags(self):
             if f == 'CXXFLAGS':
                 self.env.append_value(f, ['-fno-rtti'])
 
+        print("<!> linkflags: ", f'--target={clang_arch}', build_util.get_target_platform())
         self.env.append_value('LINKFLAGS', [f'--target={clang_arch}'])
 
     elif "macos" == build_util.get_target_os():
@@ -1449,6 +1454,11 @@ def run_tests(ctx, valgrind = False, configfile = None):
             if 'web' in ctx.env.PLATFORM: # should be moved to TEST_LAUNCH_ARGS
                 cmd = '%s %s' % (ctx.env['NODEJS'][0], cmd)
 
+            if ctx.env.PLATFORM != ctx.env.BUILD_PLATFORM and 'arm64-linux' in ctx.env.PLATFORM:
+                cmd = '%s %s' % ('qemu-aarch64 -L /usr/aarch64-linux-gnu', cmd) # TODO: move to ctx.env['QEMU_AARCH64']
+
+        print("<!> run_tests: ", cmd, "platform", ctx.env.PLATFORM, "build_platform", ctx.env.BUILD_PLATFORM)
+
         # disable shortly during beta release, due to issue with jctest + test_gui
         valgrind = False
         if valgrind:
@@ -1596,7 +1606,7 @@ def detect(conf):
         print ("Codesign disabled", Options.options.skip_codesign)
 
     # Vulkan support
-    if Options.options.with_vulkan and build_util.get_target_platform() in ('x86_64-ios','js-web','wasm-web'):
+    if Options.options.with_vulkan and build_util.get_target_platform() in ('arm64-linux', 'x86_64-ios', 'js-web', 'wasm-web'):
         conf.fatal('Vulkan is unsupported on %s' % build_util.get_target_platform())
 
     if 'win32' in platform:
@@ -1731,6 +1741,11 @@ def detect(conf):
             conf.env.CC = "clang"
             conf.env.CPP = "clang -E"
 
+        if platform == 'arm64-linux' and platform != host_platform and not Options.options.skip_tests:
+            # We'll use qemu to run tests on non-arm64 platforms
+            if not conf.env['QEMU_AARCH64']:
+                conf.find_program('qemu-aarch64', var='QEMU_AARCH64', mandatory = True)
+
     platform_setup_tools(conf, build_util)
 
     # jg: this whole thing is a 'dirty hack' to be able to pick up our own SDKs
@@ -1773,9 +1788,11 @@ def detect(conf):
         if not 'CLANG' in conf.env:
             conf.find_program('clang',   var='CLANG', mandatory = True)
             os.environ['CLANG'] = conf.env.CLANG[0]
+            print("<!> detect api gen: os.environ['CLANG']", os.environ['CLANG'])
         if not 'CLANGPP' in conf.env:
             conf.find_program('clang++', var='CLANGPP', mandatory = True)
             os.environ['CLANGPP'] = conf.env.CLANGPP[0]
+            print("<!> detect api gen: os.environ['CLANGPP']", os.environ['CLANGPP'])
 
     # NOTE: We override after check_tool. Otherwise waf gets confused and CXX_NAME etc are missing..
     if platform in ('js-web', 'wasm-web'):
@@ -1881,8 +1898,9 @@ def detect(conf):
         conf.env['LIB_OPENGL'] = ['EGL', 'GLESv1_CM', 'GLESv2']
     elif platform in ('win32', 'x86_64-win32'):
         conf.env['LINKFLAGS_OPENGL'] = ['opengl32.lib', 'glu32.lib']
-    elif platform in ('x86_64-linux','arm64-linux'):
+    elif platform in ('x86_64-linux', 'arm64-linux'):
         conf.env['LIB_OPENGL'] = ['GL', 'GLU']
+        conf.env['LIB_OPENGLES'] = ['EGL', 'GLESv1_CM', 'GLESv2']
 
     if platform in ('x86_64-macos','arm64-macos'):
         conf.env['FRAMEWORK_OPENAL'] = ['OpenAL']
@@ -1917,6 +1935,7 @@ def detect(conf):
     conf.env['STLIB_RECORD_NULL'] = 'record_null'
 
     conf.env['STLIB_GRAPHICS']          = ['graphics', 'graphics_transcoder_basisu', 'basis_transcoder']
+    conf.env['STLIB_GRAPHICS_OPENGLES'] = ['graphics_opengles', 'graphics_transcoder_basisu', 'basis_transcoder']
     conf.env['STLIB_GRAPHICS_VULKAN']   = ['graphics_vulkan', 'graphics_transcoder_basisu', 'basis_transcoder']
     conf.env['STLIB_GRAPHICS_WEBGPU']   = ['graphics_webgpu', 'graphics_transcoder_basisu', 'basis_transcoder']
     conf.env['STLIB_GRAPHICS_NULL']     = ['graphics_null', 'graphics_transcoder_null']
diff --git a/ci/ci.py b/ci/ci.py
index 468b4ffeee..5d12f11abb 100644
--- a/ci/ci.py
+++ b/ci/ci.py
@@ -244,8 +244,8 @@ def build_engine(platform, channel, with_valgrind = False, with_asan = False, wi
     if with_vanilla_lua:
         waf_opts.append('--use-vanilla-lua')
 
-    if platform == 'x86_64-linux':
-        args.append('build_sdk_headers') # gather headers after a successful build
+    # if platform == 'x86_64-linux':
+    #     args.append('build_sdk_headers') # gather headers after a successful build
 
     cmd = ' '.join(args + opts)
 
@@ -375,7 +375,7 @@ def install_ext(platform = None):
     call("python scripts/build.py install_ext %s" % ' '.join(opts))
 
 def build_bob(channel, branch = None):
-    args = "python scripts/build.py install_sdk install_ext sync_archive build_bob archive_bob".split()
+    args = "python scripts/build.py install_sdk install_ext build_bob archive_bob --skip-tests".split()
     opts = []
     opts.append("--channel=%s" % channel)
 
diff --git a/com.dynamo.cr/com.dynamo.cr.bob/scripts/copy_from_bob.sh b/com.dynamo.cr/com.dynamo.cr.bob/scripts/copy_from_bob.sh
new file mode 100644
index 0000000000..dd537da329
--- /dev/null
+++ b/com.dynamo.cr/com.dynamo.cr.bob/scripts/copy_from_bob.sh
@@ -0,0 +1,145 @@
+# Copyright 2020-2024 The Defold Foundation
+# Copyright 2014-2020 King
+# Copyright 2009-2014 Ragnar Svensson, Christian Murray
+# Licensed under the Defold License version 1.0 (the "License"); you may not use
+# this file except in compliance with the License.
+# 
+# You may obtain a copy of the License, together with FAQs at
+# https://www.defold.com/license
+# 
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+# NOTE: This script is only used for CI
+# The corresponding file for development is build.xml
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+set -e
+
+SHA1=`git log --pretty=%H -n1`
+
+# DEBUG
+# DYNAMO_HOME=tmp
+# rm -rf tmp
+# mkdir -p tmp
+
+back_cp () {
+    if [[ `unzip -Z1 $DYNAMO_HOME/share/java/bob.jar | grep "$2"` ]];then
+        echo "back_cp: extracting '$2' to '$1'"
+        mkdir -p $(dirname $1)
+        unzip -p $DYNAMO_HOME/share/java/bob.jar "$2" > "$1"
+    else
+        echo "back_cp: '$2' does not exist"
+    fi
+}
+
+# SPIRV toolchain
+back_cp $DYNAMO_HOME/ext/bin/x86_64-macos/glslang libexec/x86_64-macos/glslang
+back_cp $DYNAMO_HOME/ext/bin/arm64-macos/glslang libexec/arm64-macos/glslang
+back_cp $DYNAMO_HOME/ext/bin/x86_64-linux/glslang libexec/x86_64-linux/glslang
+back_cp $DYNAMO_HOME/ext/bin/arm64-linux/glslang libexec/arm64-linux/glslang
+back_cp $DYNAMO_HOME/ext/bin/x86_64-win32/glslang.exe libexec/x86_64-win32/glslang.exe
+
+back_cp $DYNAMO_HOME/ext/bin/x86_64-macos/tint libexec/x86_64-macos/tint
+back_cp $DYNAMO_HOME/ext/bin/arm64-macos/tint libexec/arm64-macos/tint
+back_cp $DYNAMO_HOME/ext/bin/x86_64-linux/tint libexec/x86_64-linux/tint
+#back_cp $DYNAMO_HOME/ext/bin/arm64-linux/tint libexec/arm64-linux/tint
+back_cp $DYNAMO_HOME/ext/bin/x86_64-win32/tint.exe libexec/x86_64-win32/tint.exe
+
+#
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/share/builtins.zip lib/builtins.zip
+
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/arm64-android/classes.dex lib/classes.dex
+back_cp $DYNAMO_HOME/ext/share/java/android.jar lib/android.jar
+
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/share/java/dlib.jar lib/dlib.jar
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/share/java/modelimporter.jar lib/modelimporter.jar
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/share/java/shaderc.jar lib/shaderc.jar
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/share/java/texturecompiler.jar lib/texturecompiler.jar
+
+# TEXC Shared
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/x86_64-linux/libtexc_shared.so lib/x86_64-linux/libtexc_shared.so
+#back_cp $DYNAMO_HOME/archive/${SHA1}/engine/arm64-linux/libtexc_shared.so lib/arm64-linux/libtexc_shared.so
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/x86_64-win32/texc_shared.dll lib/x86_64-win32/texc_shared.dll
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/x86_64-macos/libtexc_shared.dylib lib/x86_64-macos/libtexc_shared.dylib
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/arm64-macos/libtexc_shared.dylib lib/arm64-macos/libtexc_shared.dylib
+
+# MODELC Shared
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/x86_64-linux/libmodelc_shared.so lib/x86_64-linux/libmodelc_shared.so
+#back_cp $DYNAMO_HOME/archive/${SHA1}/engine/arm64-linux/libmodelc_shared.so lib/arm64-linux/libmodelc_shared.so
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/x86_64-win32/modelc_shared.dll lib/x86_64-win32/modelc_shared.dll
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/x86_64-macos/libmodelc_shared.dylib lib/x86_64-macos/libmodelc_shared.dylib
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/arm64-macos/libmodelc_shared.dylib lib/arm64-macos/libmodelc_shared.dylib
+
+# SHADERC Shared
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/x86_64-linux/libshaderc_shared.so lib/x86_64-linux/libshaderc_shared.so
+#back_cp $DYNAMO_HOME/archive/${SHA1}/engine/arm64-linux/libshaderc_shared.so lib/arm64-linux/libshaderc_shared.so
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/x86_64-win32/shaderc_shared.dll lib/x86_64-win32/shaderc_shared.dll
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/x86_64-macos/libshaderc_shared.dylib lib/x86_64-macos/libshaderc_shared.dylib
+back_cp $DYNAMO_HOME/archive/${SHA1}/engine/arm64-macos/libshaderc_shared.dylib lib/arm64-macos/libshaderc_shared.dylib
+
+
+back_cp $DYNAMO_HOME/ext/bin/x86_64-linux/luajit-32 libexec/x86_64-linux/luajit-32
+back_cp $DYNAMO_HOME/ext/bin/x86_64-linux/luajit-64 libexec/x86_64-linux/luajit-64
+back_cp $DYNAMO_HOME/ext/bin/arm64-linux/luajit-32 libexec/arm64-linux/luajit-32
+back_cp $DYNAMO_HOME/ext/bin/arm64-linux/luajit-64 libexec/arm64-linux/luajit-64
+back_cp $DYNAMO_HOME/ext/bin/x86_64-macos/luajit-64 libexec/x86_64-macos/luajit-64
+back_cp $DYNAMO_HOME/ext/bin/arm64-macos/luajit-64 libexec/arm64-macos/luajit-64
+back_cp $DYNAMO_HOME/ext/bin/win32/luajit-32.exe libexec/x86_64-win32/luajit-32.exe
+back_cp $DYNAMO_HOME/ext/bin/x86_64-win32/luajit-64.exe libexec/x86_64-win32/luajit-64.exe
+back_cp $DYNAMO_HOME/ext/share/luajit-share.zip lib/luajit-share.zip
+unzip -o $DYNAMO_HOME/ext/share/luajit-share.zip -d $DYNAMO_HOME/ext/share/
+# jar cfM lib/luajit-share.zip -C $DYNAMO_HOME/ext/share/ luajit
+
+#LIPO
+back_cp $DYNAMO_HOME/ext/bin/x86_64-macos/lipo libexec/x86_64-macos/lipo
+back_cp $DYNAMO_HOME/ext/bin/arm64-macos/lipo libexec/arm64-macos/lipo
+back_cp $DYNAMO_HOME/ext/bin/x86_64-linux/lipo libexec/x86_64-linux/lipo
+#back_cp $DYNAMO_HOME/ext/bin/arm64-linux/lipo libexec/arm64-linux/lipo
+back_cp $DYNAMO_HOME/ext/bin/x86_64-win32/lipo.exe libexec/x86_64-win32/lipo.exe
+
+copy () {
+    back_cp $DYNAMO_HOME/archive/${SHA1}/engine/$1 libexec/$2
+}
+
+copy x86_64-linux/stripped/dmengine x86_64-linux/dmengine
+copy x86_64-linux/stripped/dmengine_release x86_64-linux/dmengine_release
+# copy x86_64-linux/stripped/dmengine_headless x86_64-linux/dmengine_headless
+#copy arm64-linux/stripped/dmengine arm64-linux/dmengine
+#copy arm64-linux/stripped/dmengine_release arm64-linux/dmengine_release
+# copy arm64-linux/stripped/dmengine_headless arm64-linux/dmengine_headless
+copy x86_64-macos/stripped/dmengine x86_64-macos/dmengine
+copy x86_64-macos/stripped/dmengine_release x86_64-macos/dmengine_release
+# copy x86_64-macos/stripped/dmengine_headless x86_64-macos/dmengine_headless
+copy arm64-macos/stripped/dmengine arm64-macos/dmengine
+copy arm64-macos/stripped/dmengine_release arm64-macos/dmengine_release
+# copy arm64-macos/stripped/dmengine_headless arm64-macos/dmengine_headless
+# copy win32/dmengine.exe x86-win32/dmengine.exe
+# copy win32/dmengine_release.exe x86-win32/dmengine_release.exe
+# copy win32/dmengine_headless.exe x86-win32/dmengine_headless.exe
+copy x86_64-win32/dmengine.exe x86_64-win32/dmengine.exe
+copy x86_64-win32/dmengine_release.exe x86_64-win32/dmengine_release.exe
+# copy x86_64-win32/dmengine_headless.exe x86_64-win32/dmengine_headless.exe
+copy arm64-ios/stripped/dmengine arm64-ios/dmengine
+copy arm64-ios/stripped/dmengine_release arm64-ios/dmengine_release
+copy x86_64-ios/stripped/dmengine x86_64-ios/dmengine
+copy x86_64-ios/stripped/dmengine_release x86_64-ios/dmengine_release
+# copy armv7-android/stripped/libdmengine.so armv7-android/libdmengine.so
+# copy armv7-android/stripped/libdmengine_release.so armv7-android/libdmengine_release.so
+copy arm64-android/stripped/libdmengine.so arm64-android/libdmengine.so # TODO only valid once arm64-android CI target is present --jbnn
+copy arm64-android/stripped/libdmengine_release.so arm64-android/libdmengine_release.so # TODO only valid once arm64-android CI target is present --jbnn
+# copy js-web/dmengine.js js-web/dmengine.js
+# copy js-web/dmengine.js.mem js-web/dmengine.js.mem
+# copy js-web/dmengine_release.js js-web/dmengine_release.js
+# copy js-web/dmengine_release.js.mem js-web/dmengine_release.js.mem
+copy wasm-web/dmengine.js wasm-web/dmengine.js
+copy wasm-web/dmengine.wasm wasm-web/dmengine.wasm
+copy wasm-web/dmengine_release.js wasm-web/dmengine_release.js
+copy wasm-web/dmengine_release.wasm wasm-web/dmengine_release.wasm
+
+if [ -e "${DIR}/copy_private.sh" ]; then
+    sh ${DIR}/copy_private.sh
+fi
diff --git a/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java b/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java
index 8070a45a1d..871427b358 100644
--- a/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java
+++ b/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java
@@ -786,7 +786,7 @@ public class Bob {
         try {
             TimeProfiler.start("ParseCommandLine");
             if (cmd.hasOption("version")) {
-                System.out.println(String.format("bob.jar version: %s  sha1: %s  built: %s", EngineVersion.version, EngineVersion.sha1, EngineVersion.timestamp));
+                System.out.println(String.format("<!> bob.jar version: %s  sha1: %s  built: %s", EngineVersion.version, EngineVersion.sha1, EngineVersion.timestamp));
                 return new InvocationResult(true, Collections.emptyList());
             }
 
diff --git a/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderCompilers.java b/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderCompilers.java
index e54b20a03e..e5e530045a 100644
--- a/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderCompilers.java
+++ b/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderCompilers.java
@@ -45,13 +45,22 @@ public class ShaderCompilers {
                 case X86Win32:
                 case X86_64Win32:
                 case X86Linux:
-                case Arm64Linux:
                 case X86_64Linux: {
                     if (shaderType == ShaderDesc.ShaderType.SHADER_TYPE_COMPUTE) {
                         shaderLanguages.add(ShaderDesc.Language.LANGUAGE_GLSL_SM430);
                     } else {
                         shaderLanguages.add(ShaderDesc.Language.LANGUAGE_GLSL_SM140);
                     }
+                    if (shaderType != ShaderDesc.ShaderType.SHADER_TYPE_COMPUTE) {
+                        shaderLanguages.add(ShaderDesc.Language.LANGUAGE_GLES_SM300);
+                        shaderLanguages.add(ShaderDesc.Language.LANGUAGE_GLES_SM100);
+                    }
+                } break;
+                case Arm64Linux: {
+                    if (shaderType != ShaderDesc.ShaderType.SHADER_TYPE_COMPUTE) {
+                        shaderLanguages.add(ShaderDesc.Language.LANGUAGE_GLES_SM300);
+                        shaderLanguages.add(ShaderDesc.Language.LANGUAGE_GLES_SM100);
+                    }
                 } break;
                 case Arm64Ios:
                 case X86_64Ios: {
diff --git a/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderProgramBuilder.java b/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderProgramBuilder.java
index eb2ef1ae25..2709d6aee5 100644
--- a/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderProgramBuilder.java
+++ b/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderProgramBuilder.java
@@ -140,6 +140,8 @@ public abstract class ShaderProgramBuilder extends Builder {
             throw new CompileExceptionError("Unknown platform for shader program '" + resourceOutputPath + "'': " + platform);
         }
 
+        System.out.println("<!> Making shader desc for " + resourceOutputPath + " with platform " + platform);
+
         String finalShaderSource                 = shaderPreprocessor.getCompiledSource();
         IShaderCompiler shaderCompiler           = project.getShaderCompiler(platformKey);
         ShaderCompileResult shaderCompilerResult = shaderCompiler.compile(finalShaderSource, shaderType, resourceOutputPath, outputSpirv, outputWGSL);
diff --git a/editor/test/resources/test_project/app_manifest/vulkan.appmanifest b/editor/test/resources/test_project/app_manifest/vulkan.appmanifest
index e1cb688935..8f3c6447de 100644
--- a/editor/test/resources/test_project/app_manifest/vulkan.appmanifest
+++ b/editor/test/resources/test_project/app_manifest/vulkan.appmanifest
@@ -31,7 +31,7 @@ platforms:
     arm64-linux:
         context:
             excludeLibs: ["graphics"]
-            excludeSymbols: ["GraphicsAdapterOpenGL"]
+            excludeSymbols: ["GraphicsAdapterOpenGLES"]
             symbols: ["GraphicsAdapterVulkan"]
             libs: ["graphics_vulkan","X11-xcb"]
             dynamicLibs: ["vulkan"]
diff --git a/engine/dlib/src/dlib/job_thread.cpp b/engine/dlib/src/dlib/job_thread.cpp
index 887a9975ad..3fb865ad08 100644
--- a/engine/dlib/src/dlib/job_thread.cpp
+++ b/engine/dlib/src/dlib/job_thread.cpp
@@ -145,6 +145,8 @@ HContext Create(const JobThreadCreationParams& create_params)
         dmSnPrintf(name_buf, sizeof(name_buf), "%s_%d", create_params.m_ThreadNames[i], i);
         context->m_Threads[i] = dmThread::New(JobThread, 0x80000, (void*)&context->m_ThreadContext, name_buf);
     }
+#else
+    dmLogInfo("<!> NO JOB THREAD");
 #endif
     return context;
 }
diff --git a/engine/dlib/src/test/wscript b/engine/dlib/src/test/wscript
index 44fd94377f..a2faf09997 100644
--- a/engine/dlib/src/test/wscript
+++ b/engine/dlib/src/test/wscript
@@ -7,7 +7,7 @@ from BuildUtility import create_build_utility
 
 def create_test(bld, name, extra_libs = None, extra_features = None, embed_source = None,
                 extra_includes = None, extra_defines = None, skip_run = False, extra_local_libs = None, exclude_libs = None, target_name = None):
-
+    print("<!> create_test: ", name)
     # Sometimes, we need to disable certain tests until they're supported
     if not platform_supports_feature(bld.env.PLATFORM, name, {}):
         print ("Skipping test '%s' due to platform missing feature" % name)
diff --git a/engine/dlib/wscript b/engine/dlib/wscript
index a64c762f4a..7220f5716f 100644
--- a/engine/dlib/wscript
+++ b/engine/dlib/wscript
@@ -50,12 +50,25 @@ def build(bld):
     TestContext.initialize_test_context(test_context, bld)
 
 def shutdown(sht):
+    import time
+    print("<!> dlib: shutdown")
+    time.sleep(1)
+
     if not TestContext.is_valid(test_context) or Options.options.skip_build_tests or Options.options.skip_tests:
+        print("<!> dlib: skip tests: ", not TestContext.is_valid(test_context), Options.options.skip_build_tests, Options.options.skip_tests)
+        # sys.exit(1)
+        return
+
+    if True:
+        print("<!> dlib: skip tests!")
         return
 
+    print("<!> dlib: run tests!")
+    time.sleep(1)
+
     # TODO: Fix support for win32
     from waflib.Logs import warn, error
-    import urllib, urllib.request, time, atexit
+    import urllib, urllib.request, atexit
 
     server_sockets = None
     bash_prefix = ''
@@ -123,6 +136,8 @@ def shutdown(sht):
     finally:
         os.remove(configfilepath)
 
+    # sys.exit(1)
+
     # The test should work on OSX, but the CI (Catalina) just crashed with no callstack (even in lldb)
     # Postponing a fix until I've updated to Catalina on my machine as well
     # if not sys.platform in ('win32', 'darwin'):
diff --git a/engine/engine/content/builtins/input/default.gamepads b/engine/engine/content/builtins/input/default.gamepads
index 11b899312d..267a64e719 100644
--- a/engine/engine/content/builtins/input/default.gamepads
+++ b/engine/engine/content/builtins/input/default.gamepads
@@ -6017,3 +6017,34 @@ driver
     map { input: GAMEPAD_GUIDE type: GAMEPAD_TYPE_BUTTON index: 12 }
     map { input: GAMEPAD_RAW type: GAMEPAD_TYPE_AXIS index: 0 }
 }
+
+driver
+{
+    device: "retrogame_joypad" # Linux handhelds internal controller name.
+    platform: "linux"
+    dead_zone: 0.2
+    map { input: GAMEPAD_LSTICK_LEFT type: GAMEPAD_TYPE_AXIS index: 0 mod { mod: GAMEPAD_MODIFIER_NEGATE } mod { mod: GAMEPAD_MODIFIER_CLAMP } }
+    map { input: GAMEPAD_LSTICK_RIGHT type: GAMEPAD_TYPE_AXIS index: 0 mod { mod: GAMEPAD_MODIFIER_CLAMP } }
+    map { input: GAMEPAD_LSTICK_DOWN type: GAMEPAD_TYPE_AXIS index: 1 mod { mod: GAMEPAD_MODIFIER_CLAMP } }
+    map { input: GAMEPAD_LSTICK_UP type: GAMEPAD_TYPE_AXIS index: 1 mod { mod: GAMEPAD_MODIFIER_NEGATE } mod { mod: GAMEPAD_MODIFIER_CLAMP } }
+    map { input: GAMEPAD_LSTICK_CLICK type: GAMEPAD_TYPE_BUTTON index: 11 }
+    map { input: GAMEPAD_RSTICK_LEFT type: GAMEPAD_TYPE_AXIS index: 2 mod { mod: GAMEPAD_MODIFIER_NEGATE } mod { mod: GAMEPAD_MODIFIER_CLAMP } }
+    map { input: GAMEPAD_RSTICK_RIGHT type: GAMEPAD_TYPE_AXIS index: 2 mod { mod: GAMEPAD_MODIFIER_CLAMP } }
+    map { input: GAMEPAD_RSTICK_DOWN type: GAMEPAD_TYPE_AXIS index: 3 mod { mod: GAMEPAD_MODIFIER_CLAMP } }
+    map { input: GAMEPAD_RSTICK_UP type: GAMEPAD_TYPE_AXIS index: 3 mod { mod: GAMEPAD_MODIFIER_NEGATE } mod { mod: GAMEPAD_MODIFIER_CLAMP } }
+    map { input: GAMEPAD_RSTICK_CLICK type: GAMEPAD_TYPE_BUTTON index: 12 }
+    map { input: GAMEPAD_LPAD_LEFT type: GAMEPAD_TYPE_BUTTON index: 15 }
+    map { input: GAMEPAD_LPAD_RIGHT type: GAMEPAD_TYPE_BUTTON index: 16 }
+    map { input: GAMEPAD_LPAD_DOWN type: GAMEPAD_TYPE_BUTTON index: 14 }
+    map { input: GAMEPAD_LPAD_UP type: GAMEPAD_TYPE_BUTTON index: 13 }
+    map { input: GAMEPAD_RPAD_LEFT type: GAMEPAD_TYPE_BUTTON index: 3 }
+    map { input: GAMEPAD_RPAD_RIGHT type: GAMEPAD_TYPE_BUTTON index: 1 }
+    map { input: GAMEPAD_RPAD_DOWN type: GAMEPAD_TYPE_BUTTON index: 0 }
+    map { input: GAMEPAD_RPAD_UP type: GAMEPAD_TYPE_BUTTON index: 2 }
+    map { input: GAMEPAD_LTRIGGER type: GAMEPAD_TYPE_BUTTON index: 6 }
+    map { input: GAMEPAD_LSHOULDER type: GAMEPAD_TYPE_BUTTON index: 4 }
+    map { input: GAMEPAD_RTRIGGER type: GAMEPAD_TYPE_BUTTON index: 7 }
+    map { input: GAMEPAD_RSHOULDER type: GAMEPAD_TYPE_BUTTON index: 5 }
+    map { input: GAMEPAD_START type: GAMEPAD_TYPE_BUTTON index: 9 }
+    map { input: GAMEPAD_BACK type: GAMEPAD_TYPE_BUTTON index: 8 }
+}
diff --git a/engine/engine/src/engine.cpp b/engine/engine/src/engine.cpp
index 3f4d32763c..80a6c0e1be 100644
--- a/engine/engine/src/engine.cpp
+++ b/engine/engine/src/engine.cpp
@@ -589,11 +589,12 @@ namespace dmEngine
     {
         switch(family)
         {
-            case dmGraphics::ADAPTER_FAMILY_NULL:   return dmPlatform::PLATFORM_GRAPHICS_API_NULL;
-            case dmGraphics::ADAPTER_FAMILY_OPENGL: return dmPlatform::PLATFORM_GRAPHICS_API_OPENGL;
-            case dmGraphics::ADAPTER_FAMILY_VULKAN: return dmPlatform::PLATFORM_GRAPHICS_API_VULKAN;
-            case dmGraphics::ADAPTER_FAMILY_VENDOR: return dmPlatform::PLATFORM_GRAPHICS_API_VENDOR;
-            case dmGraphics::ADAPTER_FAMILY_WEBGPU: return dmPlatform::PLATFORM_GRAPHICS_API_WEBGPU;
+            case dmGraphics::ADAPTER_FAMILY_NULL:     return dmPlatform::PLATFORM_GRAPHICS_API_NULL;
+            case dmGraphics::ADAPTER_FAMILY_OPENGL:   return dmPlatform::PLATFORM_GRAPHICS_API_OPENGL;
+            case dmGraphics::ADAPTER_FAMILY_OPENGLES: return dmPlatform::PLATFORM_GRAPHICS_API_OPENGLES;
+            case dmGraphics::ADAPTER_FAMILY_VULKAN:   return dmPlatform::PLATFORM_GRAPHICS_API_VULKAN;
+            case dmGraphics::ADAPTER_FAMILY_VENDOR:   return dmPlatform::PLATFORM_GRAPHICS_API_VENDOR;
+            case dmGraphics::ADAPTER_FAMILY_WEBGPU:   return dmPlatform::PLATFORM_GRAPHICS_API_WEBGPU;
             default:break;
         }
         assert(0);
diff --git a/engine/engine/src/wscript b/engine/engine/src/wscript
index b67dbe7530..c8d22e462f 100755
--- a/engine/engine/src/wscript
+++ b/engine/engine/src/wscript
@@ -77,6 +77,10 @@ def build(bld):
     graphics_lib = 'GRAPHICS DMGLFW OPENGL'
     graphics_lib_symbols = ['GraphicsAdapterOpenGL']
 
+    if bld.env['PLATFORM'] in ('arm64-linux'):
+        graphics_lib = 'GRAPHICS_OPENGLES DMGLFW OPENGLES'
+        graphics_lib_symbols = ['GraphicsAdapterOpenGLES']
+
     if waflib.Options.options.with_vulkan:
         graphics_lib += ' GRAPHICS_VULKAN VULKAN'
         graphics_lib_symbols.append('GraphicsAdapterVulkan')
diff --git a/engine/engine/wscript b/engine/engine/wscript
index e002b73190..58e0f6aaf5 100644
--- a/engine/engine/wscript
+++ b/engine/engine/wscript
@@ -108,6 +108,13 @@ def configure(conf):
     elif target_os == "win":
         conf.env.append_value('LINKFLAGS', ['opengl32.lib', 'user32.lib', 'shell32.lib', 'xinput9_1_0.lib', 'dbghelp.lib', 'dinput8.lib', 'dxguid.lib'])
 
+        # ANGLE
+        # conf.env.append_value('LINKFLAGS', ['libEGL.dll.lib', 'libGLESv1_CM.dll.lib', 'libGLESv2.dll.lib', 'user32.lib', 'shell32.lib', 'xinput9_1_0.lib', 'dbghelp.lib', 'dinput8.lib', 'dxguid.lib'])
+
+        # print("<!> engine: ANGLE LIBPATH AND INCLUDES + TEST PATH %s" % (build_util.get_dynamo_ext("include", "AL")))
+        # conf.env.append_value('INCLUDES', 'D:\\a\\ci-test\\ci-test\\patches\\angle\\include')
+        # conf.env.append_value('LIBPATH', 'D:\\a\\ci-test\\ci-test\\patches\\angle\\out\\Debug')
+
         # For Vulkan
         if waflib.Options.options.with_vulkan:
             conf.env.append_value('LINKFLAGS', ['vulkan-1.lib'])
diff --git a/engine/gamesys/src/gamesys/components/comp_model.cpp b/engine/gamesys/src/gamesys/components/comp_model.cpp
index 34afa2123c..559eed9ca6 100644
--- a/engine/gamesys/src/gamesys/components/comp_model.cpp
+++ b/engine/gamesys/src/gamesys/components/comp_model.cpp
@@ -171,16 +171,17 @@ namespace dmGameSystem
         world->m_Components.SetCapacity(comp_count);
         world->m_RenderObjects.SetCapacity(comp_count);
         // position, normal, tangent, color, texcoord0, texcoord1 * sizeof(float)
-        DM_STATIC_ASSERT( sizeof(dmRig::RigModelVertex) == ((3+3+4+4+2+2)*4), Invalid_Struct_Size);
+        // DM_STATIC_ASSERT( sizeof(dmRig::RigModelVertex) == ((3*4+3*4+4*2+4*2+2*2+2*2)), Invalid_Struct_Size);
+        DM_STATIC_ASSERT( sizeof(dmRig::RigModelVertex) == ((3*4+4*1+2*2)), Invalid_Struct_Size);
 
         dmGraphics::HContext graphics_context = dmRender::GetGraphicsContext(render_context);
         dmGraphics::HVertexStreamDeclaration stream_declaration_vertex = dmGraphics::NewVertexStreamDeclaration(graphics_context);
         dmGraphics::AddVertexStream(stream_declaration_vertex, "position",  3, dmGraphics::TYPE_FLOAT, false);
-        dmGraphics::AddVertexStream(stream_declaration_vertex, "normal",    3, dmGraphics::TYPE_FLOAT, false);
-        dmGraphics::AddVertexStream(stream_declaration_vertex, "tangent",   4, dmGraphics::TYPE_FLOAT, false);
-        dmGraphics::AddVertexStream(stream_declaration_vertex, "color",     4, dmGraphics::TYPE_FLOAT, false);
-        dmGraphics::AddVertexStream(stream_declaration_vertex, "texcoord0", 2, dmGraphics::TYPE_FLOAT, false);
-        dmGraphics::AddVertexStream(stream_declaration_vertex, "texcoord1", 2, dmGraphics::TYPE_FLOAT, false);
+        dmGraphics::AddVertexStream(stream_declaration_vertex, "normal",    4, dmGraphics::TYPE_BYTE, true);
+        // dmGraphics::AddVertexStream(stream_declaration_vertex, "tangent",   4, dmGraphics::TYPE_SHORT, true);
+        // dmGraphics::AddVertexStream(stream_declaration_vertex, "color",     4, dmGraphics::TYPE_SHORT, true);
+        dmGraphics::AddVertexStream(stream_declaration_vertex, "texcoord0", 2, dmGraphics::TYPE_SHORT, true);
+        // dmGraphics::AddVertexStream(stream_declaration_vertex, "texcoord1", 2, dmGraphics::TYPE_SHORT, true);
 
         dmGraphics::HVertexStreamDeclaration stream_declaration_instance = dmGraphics::NewVertexStreamDeclaration(graphics_context, dmGraphics::VERTEX_STEP_FUNCTION_INSTANCE);
         dmGraphics::AddVertexStream(stream_declaration_instance, "mtx_world",  16, dmGraphics::TYPE_FLOAT, false);
diff --git a/engine/gamesys/src/gamesys/resources/res_model.cpp b/engine/gamesys/src/gamesys/resources/res_model.cpp
index b6f1599089..e9c60bec49 100644
--- a/engine/gamesys/src/gamesys/resources/res_model.cpp
+++ b/engine/gamesys/src/gamesys/resources/res_model.cpp
@@ -114,19 +114,25 @@ namespace dmGameSystem
             for (int c = 0; c < 3; ++c)
             {
                 out_write_ptr->pos[c] = *positions++;
-                out_write_ptr->normal[c] = normals ? *normals++ : 0.0f;
+                // out_write_ptr->normal[c] = normals ? *normals++ : 0;
             }
 
+            for (int c = 0; c < 3; ++c)
+            {
+                out_write_ptr->normal[c] = normals ? (int8_t)((*normals++) * 127.0f) : 0;
+            }
+            out_write_ptr->normal[3] = (int8_t)(1.0f * 127.0f);
+
             for (int c = 0; c < 4; ++c)
             {
-                out_write_ptr->color[c] = colors ? *colors++ : 1.0f;
-                out_write_ptr->tangent[c] = tangents ? *tangents++ : 0.0f;
+                // out_write_ptr->color[c] = colors ? (int16_t)((*colors++) * 32767.0f) : 1;
+                // out_write_ptr->tangent[c] = tangents ? (int16_t)((*tangents++) * 32767.0f) : 0;
             }
 
             for (int c = 0; c < 2; ++c)
             {
-                out_write_ptr->uv0[c] = uv0 ? *uv0++ : 0.0f;
-                out_write_ptr->uv1[c] = uv1 ? *uv1++ : 0.0f;
+                out_write_ptr->uv0[c] = uv0 ? (int16_t)((*uv0++) * 32767.0f) : 0;
+                // out_write_ptr->uv1[c] = uv1 ? (int16_t)((*uv1++) * 32767.0f) : 0;
             }
 
             out_write_ptr++;
diff --git a/engine/gamesys/wscript b/engine/gamesys/wscript
index ad03b4201c..fddc0a3fe7 100644
--- a/engine/gamesys/wscript
+++ b/engine/gamesys/wscript
@@ -40,8 +40,12 @@ def configure(conf):
         pass
     elif re.match('arm.*?android', platform):
         conf.env.append_value('LINKFLAGS', ['-lEGL', '-lGLESv1_CM', '-lGLESv2', '-landroid'])
-    elif platform == "linux" or platform == "x86_64-linux":
-        conf.env['LIB_X'] = ['Xext', 'X11', 'Xi', 'GL', 'GLU']
+    elif platform == "x86_64-linux":
+        print("<!> gamesys: GL LINKFLAGS")
+        # conf.env['LIB_X'] = ['Xext', 'X11', 'Xi', 'GL', 'GLU']
+    elif platform == "arm64-linux":
+        print("<!> gamesys: GLES LINKFLAGS")
+        # conf.env['LIB_X'] = ['Xext', 'X11', 'Xi', 'EGL', 'GLESv1_CM', 'GLESv2']
     elif "win32" in platform:
         conf.env.append_value('LINKFLAGS', ['opengl32.lib', 'user32.lib', 'shell32.lib', 'xinput9_1_0.lib', 'dinput8.lib', 'dxguid.lib'])
 
diff --git a/engine/graphics/src/graphics.cpp b/engine/graphics/src/graphics.cpp
index 13a1b2f1a5..8a708dc78c 100644
--- a/engine/graphics/src/graphics.cpp
+++ b/engine/graphics/src/graphics.cpp
@@ -106,6 +106,8 @@ namespace dmGraphics
             return ADAPTER_FAMILY_NULL;
         if (dmStrCaseCmp("opengl", adapter_name) == 0)
             return ADAPTER_FAMILY_OPENGL;
+        if (dmStrCaseCmp("opengles", adapter_name) == 0)
+            return ADAPTER_FAMILY_OPENGLES;
         if (dmStrCaseCmp("vulkan", adapter_name) == 0)
             return ADAPTER_FAMILY_VULKAN;
         if (dmStrCaseCmp("webgpu", adapter_name) == 0)
@@ -125,6 +127,7 @@ namespace dmGraphics
             GRAPHICS_ENUM_TO_STR_CASE(ADAPTER_FAMILY_NONE);
             GRAPHICS_ENUM_TO_STR_CASE(ADAPTER_FAMILY_NULL);
             GRAPHICS_ENUM_TO_STR_CASE(ADAPTER_FAMILY_OPENGL);
+            GRAPHICS_ENUM_TO_STR_CASE(ADAPTER_FAMILY_OPENGLES);
             GRAPHICS_ENUM_TO_STR_CASE(ADAPTER_FAMILY_VULKAN);
             GRAPHICS_ENUM_TO_STR_CASE(ADAPTER_FAMILY_VENDOR);
             GRAPHICS_ENUM_TO_STR_CASE(ADAPTER_FAMILY_WEBGPU);
@@ -318,9 +321,11 @@ namespace dmGraphics
         assert(shader_desc);
         ShaderDesc::Shader* selected_shader = 0x0;
 
+        dmLogInfo("<!> GetShaderProgram: shader_desc->m_Shaders.m_Count: %d", shader_desc->m_Shaders.m_Count);
         for(uint32_t i = 0; i < shader_desc->m_Shaders.m_Count; ++i)
         {
             ShaderDesc::Shader* shader = &shader_desc->m_Shaders.m_Data[i];
+            dmLogInfo("<!> GetShaderProgram: shader->m_Language: %d, shader_desc->m_ShaderType: %d", shader->m_Language, shader_desc->m_ShaderType);
             if(IsShaderLanguageSupported(context, shader->m_Language, shader_desc->m_ShaderType))
             {
                 if (shader->m_VariantTextureArray)
diff --git a/engine/graphics/src/graphics.h b/engine/graphics/src/graphics.h
index d90b5da552..748c28582c 100644
--- a/engine/graphics/src/graphics.h
+++ b/engine/graphics/src/graphics.h
@@ -79,12 +79,13 @@ namespace dmGraphics
 
     enum AdapterFamily
     {
-        ADAPTER_FAMILY_NONE   = -1,
-        ADAPTER_FAMILY_NULL   = 1,
-        ADAPTER_FAMILY_OPENGL = 2,
-        ADAPTER_FAMILY_VULKAN = 3,
-        ADAPTER_FAMILY_VENDOR = 4,
-        ADAPTER_FAMILY_WEBGPU = 5,
+        ADAPTER_FAMILY_NONE     = -1,
+        ADAPTER_FAMILY_NULL     = 1,
+        ADAPTER_FAMILY_OPENGL   = 2,
+        ADAPTER_FAMILY_OPENGLES = 3,
+        ADAPTER_FAMILY_VULKAN   = 4,
+        ADAPTER_FAMILY_VENDOR   = 5,
+        ADAPTER_FAMILY_WEBGPU   = 6,
     };
 
     enum AssetType
diff --git a/engine/graphics/src/null/graphics_null.cpp b/engine/graphics/src/null/graphics_null.cpp
index 085954f44e..8c93ac342a 100644
--- a/engine/graphics/src/null/graphics_null.cpp
+++ b/engine/graphics/src/null/graphics_null.cpp
@@ -910,10 +910,12 @@ namespace dmGraphics
         {
             case ShaderDesc::SHADER_TYPE_VERTEX:
                 return language == ShaderDesc::LANGUAGE_GLSL_SM140 ||
-                       language == ShaderDesc::LANGUAGE_GLSL_SM330;
+                       language == ShaderDesc::LANGUAGE_GLSL_SM330 ||
+                       language == ShaderDesc::LANGUAGE_GLES_SM300;
             case ShaderDesc::SHADER_TYPE_FRAGMENT:
                 return language == ShaderDesc::LANGUAGE_GLSL_SM140 ||
-                       language == ShaderDesc::LANGUAGE_GLSL_SM330;
+                       language == ShaderDesc::LANGUAGE_GLSL_SM330 ||
+                       language == ShaderDesc::LANGUAGE_GLES_SM300;
             case ShaderDesc::SHADER_TYPE_COMPUTE:
                 return language == ShaderDesc::LANGUAGE_GLSL_SM430 ||
                        language == ShaderDesc::LANGUAGE_SPIRV;
diff --git a/engine/graphics/src/opengl/graphics_opengl.cpp b/engine/graphics/src/opengl/graphics_opengl.cpp
index 5701284911..b79abf1212 100644
--- a/engine/graphics/src/opengl/graphics_opengl.cpp
+++ b/engine/graphics/src/opengl/graphics_opengl.cpp
@@ -71,9 +71,7 @@
     #endif
 #endif
 
-#if defined(__linux__) && !defined(ANDROID)
-    #include <GL/glext.h>
-#elif defined (ANDROID)
+#if defined (ANDROID)
     #define GL_GLEXT_PROTOTYPES
     #include <GLES2/gl2ext.h>
 
@@ -83,6 +81,8 @@
 
 #elif defined (__MACH__)
     // NOP
+#elif defined (__linux__)
+    // NOP
 #elif defined (_WIN32)
     #ifdef GL_GLEXT_PROTOTYPES
         #undef GL_GLEXT_PROTOTYPES
@@ -411,10 +411,13 @@ static void LogFrameBufferError(GLenum status)
     static bool                         OpenGLIsSupported();
     static HContext                     OpenGLGetContext();
     static int8_t          g_null_adapter_priority = 1;
-    static GraphicsAdapter g_opengl_adapter(ADAPTER_FAMILY_OPENGL);
-
+	#if defined(DM_GRAPHICS_USE_OPENGLES)
+	static GraphicsAdapter g_opengl_adapter(ADAPTER_FAMILY_OPENGLES);
+    DM_REGISTER_GRAPHICS_ADAPTER(GraphicsAdapterOpenGLES, &g_opengl_adapter, OpenGLIsSupported, OpenGLRegisterFunctionTable, OpenGLGetContext, g_null_adapter_priority);
+	#else
+	static GraphicsAdapter g_opengl_adapter(ADAPTER_FAMILY_OPENGL);
     DM_REGISTER_GRAPHICS_ADAPTER(GraphicsAdapterOpenGL, &g_opengl_adapter, OpenGLIsSupported, OpenGLRegisterFunctionTable, OpenGLGetContext, g_null_adapter_priority);
-
+    #endif
     static void PostDeleteTextures(OpenGLContext*, bool);
     static bool OpenGLInitialize(HContext context, const ContextParams& params);
 
@@ -941,7 +944,11 @@ static void LogFrameBufferError(GLenum status)
 
     static void OpenGLPrintDeviceInfo(HContext context)
     {
+        #if defined(DM_GRAPHICS_USE_OPENGLES)
+        dmLogInfo("Device: OpenGL ES");
+        #else
         dmLogInfo("Device: OpenGL");
+        #endif
         dmLogInfo("Renderer: %s", (char *) glGetString(GL_RENDERER));
         dmLogInfo("Version: %s", (char *) glGetString(GL_VERSION));
         dmLogInfo("Vendor: %s", (char *) glGetString(GL_VENDOR));
@@ -1073,7 +1080,7 @@ static void LogFrameBufferError(GLenum status)
         context->m_IsGles3Version = 1; // 0 == gles 2, 1 == gles 3
         context->m_PipelineState  = GetDefaultPipelineState();
 
-#if defined(__EMSCRIPTEN__) || defined(__ANDROID__)
+#if defined(__EMSCRIPTEN__) || defined(__ANDROID__) || defined(DM_GRAPHICS_USE_OPENGLES)
         context->m_IsShaderLanguageGles = 1;
 
         const char* version = (char *) glGetString(GL_VERSION);
@@ -1147,7 +1154,6 @@ static void LogFrameBufferError(GLenum status)
             (void) SetFrontProcess( &psn );
 #endif
 
-
     #if !(defined(__EMSCRIPTEN__) || defined(GL_ES_VERSION_2_0))
         GLint n;
         glGetIntegerv(GL_NUM_EXTENSIONS, &n);
@@ -1443,7 +1449,7 @@ static void LogFrameBufferError(GLenum status)
         #endif
         }
 
-#if defined(__ANDROID__) || defined(__arm__) || defined(__arm64__) || defined(__EMSCRIPTEN__)
+#if defined(__ANDROID__) || defined(__arm__) || defined(__arm64__) || defined(__EMSCRIPTEN__) || defined(DM_GRAPHICS_USE_OPENGLES)
         if (OpenGLIsExtensionSupported(context, "GL_OES_element_index_uint") ||
             OpenGLIsExtensionSupported(context, "OES_element_index_uint"))
         {
diff --git a/engine/graphics/src/opengl/graphics_opengl_defines.h b/engine/graphics/src/opengl/graphics_opengl_defines.h
index dfde827c3a..3fe45e5bab 100644
--- a/engine/graphics/src/opengl/graphics_opengl_defines.h
+++ b/engine/graphics/src/opengl/graphics_opengl_defines.h
@@ -17,10 +17,17 @@
 
 #if defined(__linux__) && !defined(ANDROID)
 
+#if defined(__aarch64__)
+#define GL_GLEXT_PROTOTYPES
+#include <GLES3/gl3.h>
+#include <GLES2/gl2ext.h>
+#define GL_BGRA GL_BGRA_EXT
+#else
 #define GL_HAS_RENDERDOC_SUPPORT
 #include <GL/gl.h>
 #include <GL/glu.h>
 #include <GL/glut.h>
+#endif
 
 #elif defined (__MACH__)
 
diff --git a/engine/graphics/src/test/test_app_graphics.cpp b/engine/graphics/src/test/test_app_graphics.cpp
index debcab564b..927e18a43d 100644
--- a/engine/graphics/src/test/test_app_graphics.cpp
+++ b/engine/graphics/src/test/test_app_graphics.cpp
@@ -378,6 +378,10 @@ static void* EngineCreate(int argc, char** argv)
     {
         window_params.m_GraphicsApi = dmPlatform::PLATFORM_GRAPHICS_API_OPENGL;
     }
+    else if (dmGraphics::GetInstalledAdapterFamily() == dmGraphics::ADAPTER_FAMILY_OPENGLES)
+    {
+        window_params.m_GraphicsApi = dmPlatform::PLATFORM_GRAPHICS_API_OPENGLES;
+    }
 
     dmPlatform::OpenWindow(engine->m_Window, window_params);
     dmPlatform::ShowWindow(engine->m_Window);
diff --git a/engine/graphics/src/test/wscript b/engine/graphics/src/test/wscript
index 95c505e40e..4bdad53a13 100755
--- a/engine/graphics/src/test/wscript
+++ b/engine/graphics/src/test/wscript
@@ -22,6 +22,10 @@ def build(bld):
         glfw_lib = ['DMGLFW']
         glfw_js = '%s/ext/lib/%s/js/library_glfw.js' % (bld.env.DYNAMO_HOME, bld.env.PLATFORM)
 
+        if platform_supports_feature(bld.env.PLATFORM, 'opengles', {}):
+            graphics_libs = ['GRAPHICS_OPENGLES', 'graphics_transcoder_basisu']
+            extra_symbols = ['GraphicsAdapterOpenGLES']
+
         if bld.env.PLATFORM in ('armv7-android', 'arm64-android'):
             extra_libs += ['OPENGL']
 
diff --git a/engine/graphics/src/wscript b/engine/graphics/src/wscript
index 70e16b6230..8aacf55561 100755
--- a/engine/graphics/src/wscript
+++ b/engine/graphics/src/wscript
@@ -56,23 +56,31 @@ def build(bld):
 
     obj.source = [x for x in obj.source if not x.name.endswith('graphics_glfw_wrappers.cpp')]
 
-    if platform_supports_feature(bld.env.PLATFORM, 'opengl', {}):
-
-        opengl_defines = []
-        opengl_platform_includes = []
-
-        if platform_supports_feature(bld.env.PLATFORM, 'opengl_compute', {}):
-            opengl_defines = ['DM_HAVE_PLATFORM_COMPUTE_SUPPORT']
-
-        if bld.env.PLATFORM in ('x86_64-win32', 'win32'):
-            opengl_platform_includes = ['opengl/win32']
-
-        obj = bld.stlib(features        = 'cxx',
-                        includes        = ['.', 'opengl'] + proto_cpp_includes + opengl_platform_includes,
-                        source          = bld.path.ant_glob(['*.cpp', 'opengl/*.cpp']) + proto_cpp_source,
-                        use             = 'DDF DLIB',
-                        defines         = opengl_defines,
-                        target          = 'graphics')
+    for graphics_feature in ['opengl', 'opengles']:
+        if platform_supports_feature(bld.env.PLATFORM, graphics_feature, {}):
+            opengl_target = 'graphics'
+            opengl_defines = ['DM_GRAPHICS_USE_OPENGL']
+            opengl_platform_includes = []
+
+            if graphics_feature == 'opengles':
+                opengl_target = 'graphics_opengles'
+                print("<!> DM_GRAPHICS_USE_OPENGLES")
+                import time
+                time.sleep(1)
+                opengl_defines = ['DM_GRAPHICS_USE_OPENGLES']
+
+            if platform_supports_feature(bld.env.PLATFORM, 'opengl_compute', {}):
+                opengl_defines.append('DM_HAVE_PLATFORM_COMPUTE_SUPPORT')
+
+            if bld.env.PLATFORM in ('x86_64-win32', 'win32'):
+                opengl_platform_includes = ['opengl/win32']
+
+            obj = bld.stlib(features        = 'cxx',
+                            includes        = ['.', 'opengl'] + proto_cpp_includes + opengl_platform_includes,
+                            source          = bld.path.ant_glob(['*.cpp', 'opengl/*.cpp']) + proto_cpp_source,
+                            use             = 'DDF DLIB',
+                            defines         = opengl_defines,
+                            target          = opengl_target)
 
     if platform_supports_feature(bld.env.PLATFORM, 'webgpu', {}):
         obj = bld.stlib(features        = 'cxx',
diff --git a/engine/graphics/wscript b/engine/graphics/wscript
index 3b8d2ca070..bcd9315a68 100644
--- a/engine/graphics/wscript
+++ b/engine/graphics/wscript
@@ -52,8 +52,12 @@ def configure(conf):
         conf.env.append_value('LINKFLAGS', ['-framework', 'Carbon', '-framework', 'OpenGL', '-framework', 'AGL', '-framework', 'IOKit', '-framework', 'CoreVideo'])
     elif platform == "x86_64-ios":
         pass
-    elif platform in ["x86_64-linux", "arm64-linux"]:
+    elif platform == "x86_64-linux":
+        print("<!> graphics: GL LINKFLAGS")
         conf.env.append_value('LINKFLAGS', ['-lXext', '-lX11', '-lXi', '-lGL', '-lGLU', '-lpthread'])
+    elif platform == "arm64-linux":
+        print("<!> graphics: GLES LINKFLAGS")
+        conf.env.append_value('LINKFLAGS', ['-lXext', '-lX11', '-lXi', '-lEGL', '-lGLESv1_CM', '-lGLESv2', '-lpthread'])
     elif "win32" in platform:
         conf.env.append_value('LINKFLAGS', ['opengl32.lib'])
 
diff --git a/engine/gui/wscript b/engine/gui/wscript
index bd80fa2574..e18c5f8494 100644
--- a/engine/gui/wscript
+++ b/engine/gui/wscript
@@ -46,8 +46,11 @@ def configure(conf):
         conf.env.append_value('LINKFLAGS', ['-framework', 'Cocoa', '-framework', 'OpenGL', '-framework', 'AGL', '-framework', 'IOKit', '-framework', 'CoreVideo'])
     if platform in ("arm64-ios",'x86_64-ios'):
         conf.env.append_value('LINKFLAGS', ['-framework', 'UIKit', '-framework', 'OpenGLES', '-framework', 'QuartzCore', '-framework', 'CoreGraphics'])
-    if platform in ("x86_64-linux",):
+    if platform in ("x86_64-linux"):
         conf.env['LIB_X'] = ['Xext', 'X11', 'Xi', 'GL', 'GLU']
+    if platform in ("arm64-linux"):
+        print("<!> graphics/gui: GLES LINKFLAGS")
+        conf.env['LIB_X'] = ['Xext', 'X11', 'Xi', 'EGL', 'GLESv1_CM', 'GLESv2']
     if 'win32' in platform:
         conf.env.append_value('LINKFLAGS', ['opengl32.lib', 'glu32.lib', 'user32.lib', 'shell32.lib', 'xinput9_1_0.lib'])
 
diff --git a/engine/hid/src/native/hid_gamepad_driver_glfw.cpp b/engine/hid/src/native/hid_gamepad_driver_glfw.cpp
index 3338d746ad..a56ea9244e 100644
--- a/engine/hid/src/native/hid_gamepad_driver_glfw.cpp
+++ b/engine/hid/src/native/hid_gamepad_driver_glfw.cpp
@@ -355,6 +355,7 @@ namespace dmHID
     static void GetGamepadDeviceNameInternal(HContext context, int glfw_id, char name[MAX_GAMEPAD_NAME_LENGTH])
     {
         const char* device_name = dmPlatform::GetJoystickDeviceName(context->m_Window, glfw_id);
+        dmLogInfo("<!> GetGamepadDeviceNameInternal: %s", device_name);
         dmStrlCpy(name, device_name, MAX_GAMEPAD_NAME_LENGTH);
     }
 
diff --git a/engine/hid/src/test/wscript b/engine/hid/src/test/wscript
index b8ea5561fe..19e1115d0e 100644
--- a/engine/hid/src/test/wscript
+++ b/engine/hid/src/test/wscript
@@ -31,6 +31,12 @@ def build(bld):
         graphics_lib += ['OPENGL']
         graphics_lib_symbols += ['GraphicsAdapterOpenGL']
 
+    if platform_supports_feature(bld.env.PLATFORM, 'opengles', {}):
+        platform_lib = ['PLATFORM']
+        graphics_lib = ['GRAPHICS_OPENGLES', 'DMGLFW']
+        graphics_lib += ['OPENGLES']
+        graphics_lib_symbols += ['GraphicsAdapterOpenGLES']
+
     if platform_supports_feature(bld.env.PLATFORM, 'vulkan', {}) and Options.options.with_vulkan:
         graphics_lib_symbols += ['GraphicsAdapterVulkan']
         graphics_lib += ['GRAPHICS_VULKAN', 'VULKAN']
diff --git a/engine/platform/src/platform_window.h b/engine/platform/src/platform_window.h
index 5bb82224c5..3983adcbfc 100644
--- a/engine/platform/src/platform_window.h
+++ b/engine/platform/src/platform_window.h
@@ -41,11 +41,12 @@ namespace dmPlatform
 
     enum PlatformGraphicsApi
     {
-        PLATFORM_GRAPHICS_API_NULL   = 0,
-        PLATFORM_GRAPHICS_API_OPENGL = 1,
-        PLATFORM_GRAPHICS_API_VULKAN = 2,
-        PLATFORM_GRAPHICS_API_VENDOR = 3,
-        PLATFORM_GRAPHICS_API_WEBGPU = 4,
+        PLATFORM_GRAPHICS_API_NULL     = 0,
+        PLATFORM_GRAPHICS_API_OPENGL   = 1,
+        PLATFORM_GRAPHICS_API_OPENGLES = 2,
+        PLATFORM_GRAPHICS_API_VULKAN   = 3,
+        PLATFORM_GRAPHICS_API_VENDOR   = 4,
+        PLATFORM_GRAPHICS_API_WEBGPU   = 5,
     };
 
     enum DeviceState
diff --git a/engine/platform/src/platform_window_glfw.cpp b/engine/platform/src/platform_window_glfw.cpp
index fe92d1744f..084cca4e49 100644
--- a/engine/platform/src/platform_window_glfw.cpp
+++ b/engine/platform/src/platform_window_glfw.cpp
@@ -276,6 +276,7 @@ namespace dmPlatform
         switch(params.m_GraphicsApi)
         {
             case PLATFORM_GRAPHICS_API_OPENGL:
+            case PLATFORM_GRAPHICS_API_OPENGLES:
                 res = OpenWindowOpenGL(window, params);
                 break;
             case PLATFORM_GRAPHICS_API_WEBGPU:
diff --git a/engine/platform/src/platform_window_glfw3.cpp b/engine/platform/src/platform_window_glfw3.cpp
index 33f5d33201..39d96e5e61 100644
--- a/engine/platform/src/platform_window_glfw3.cpp
+++ b/engine/platform/src/platform_window_glfw3.cpp
@@ -43,7 +43,7 @@ namespace dmPlatform
 
     static void OnError(int error, const char* description)
     {
-        dmLogError("GLFW Error: %s\n", description);
+        dmLogError("GLFW Error: %s", description);
     }
 
     static void OnWindowResize(GLFWwindow* glfw_window, int width, int height)
@@ -142,26 +142,37 @@ namespace dmPlatform
 
     HWindow NewWindow()
     {
+        glfwSetErrorCallback(OnError);
+
     #ifdef __MACH__
         glfwInitHint(GLFW_COCOA_CHDIR_RESOURCES, GLFW_FALSE);
     #endif
 
     #if defined(__linux__) && !defined(ANDROID)
-        glfwInitHint(GLFW_WAYLAND_LIBDECOR, GLFW_WAYLAND_PREFER_LIBDECOR);
         glfwInitHint(GLFW_PLATFORM, GLFW_PLATFORM_X11);
-    #endif
-
+        if (glfwInit() == GL_FALSE)
+        {
+            glfwInitHint(GLFW_PLATFORM, GLFW_PLATFORM_WAYLAND);
+            // TEMPORARY HACK
+            glfwInitHint(GLFW_WAYLAND_LIBDECOR, GLFW_WAYLAND_DISABLE_LIBDECOR);
+            //
+            if (glfwInit() == GL_FALSE)
+            {
+                dmLogError("Could not initialize glfw.");
+                return 0;
+            }
+        }
+    #else
         if (glfwInit() == GL_FALSE)
         {
             dmLogError("Could not initialize glfw.");
             return 0;
         }
+    #endif
 
         dmWindow* wnd = new dmWindow;
         memset(wnd, 0, sizeof(dmWindow));
 
-        glfwSetErrorCallback(OnError);
-
         // Reset context
         memset(&g_GLFW3Context, 0, sizeof(PlatformContext));
 
@@ -181,46 +192,58 @@ namespace dmPlatform
 
     static PlatformResult OpenWindowOpenGL(dmWindow* wnd, const WindowParams& params)
     {
-        uint32_t major = 3, minor = 3;
-        if (!OpenGLGetVersion(params.m_OpenGLVersionHint, &major, &minor))
-        {
-            dmLogWarning("OpenGL version hint %d is not supported. Using default version (%d.%d)",
-                params.m_OpenGLVersionHint, major, minor);
-        }
+        glfwWindowHint(GLFW_SAMPLES, params.m_Samples);
 
-        bool use_highest_version = major == 0 && minor == 0;
-        if (!use_highest_version)
+        GLFWmonitor* fullscreen_monitor = NULL;
+        if (params.m_Fullscreen)
         {
-            glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, major);
-            glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, minor);
+            fullscreen_monitor = glfwGetPrimaryMonitor();
         }
 
-        if (params.m_OpenGLUseCoreProfileHint)
+        if (params.m_GraphicsApi == PLATFORM_GRAPHICS_API_OPENGLES)
         {
-            bool can_set_profile_and_fc = true;
+            glfwWindowHint(GLFW_CLIENT_API, GLFW_OPENGL_ES_API);
+            glfwWindowHint(GLFW_CONTEXT_CREATION_API, GLFW_EGL_CONTEXT_API);
+            glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 2);
+            glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);
 
-        #ifdef _WIN32
-            // Not supported on windows when requesting the default OpenGL version (which will equate to the highest available)
-            can_set_profile_and_fc = !use_highest_version;
-        #endif
+            wnd->m_Window = glfwCreateWindow(params.m_Width, params.m_Height, params.m_Title, fullscreen_monitor, NULL);
+        }
+        else
+        {
+            uint32_t major = 3, minor = 3;
+            if (!OpenGLGetVersion(params.m_OpenGLVersionHint, &major, &minor))
+            {
+                dmLogWarning("OpenGL version hint %d is not supported. Using default version (%d.%d)",
+                    params.m_OpenGLVersionHint, major, minor);
+            }
 
-            if (can_set_profile_and_fc)
+            bool use_highest_version = major == 0 && minor == 0;
+            if (!use_highest_version)
             {
-                glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
-                glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
+                glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, major);
+                glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, minor);
             }
-        }
 
-        glfwWindowHint(GLFW_SAMPLES, params.m_Samples);
+            if (params.m_OpenGLUseCoreProfileHint)
+            {
+                bool can_set_profile_and_fc = true;
+
+            #ifdef _WIN32
+                // Not supported on windows when requesting the default OpenGL version (which will equate to the highest available)
+                can_set_profile_and_fc = !use_highest_version;
+            #endif
+
+                if (can_set_profile_and_fc)
+                {
+                    glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
+                    glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
+                }
+            }
 
-        GLFWmonitor* fullscreen_monitor = NULL;
-        if (params.m_Fullscreen)
-        {
-            fullscreen_monitor = glfwGetPrimaryMonitor();
+            wnd->m_Window = glfwCreateWindow(params.m_Width, params.m_Height, params.m_Title, fullscreen_monitor, NULL);
         }
 
-        wnd->m_Window = glfwCreateWindow(params.m_Width, params.m_Height, params.m_Title, fullscreen_monitor, NULL);
-
         if (!wnd->m_Window)
         {
             return PLATFORM_RESULT_WINDOW_OPEN_ERROR;
@@ -282,6 +305,7 @@ namespace dmPlatform
         switch(params.m_GraphicsApi)
         {
             case PLATFORM_GRAPHICS_API_OPENGL:
+            case PLATFORM_GRAPHICS_API_OPENGLES:
                 res = OpenWindowOpenGL(window, params);
                 break;
             case PLATFORM_GRAPHICS_API_VULKAN:
diff --git a/engine/rig/src/dmsdk/rig/rig.h b/engine/rig/src/dmsdk/rig/rig.h
index 92d2a75e73..1930a4f27c 100644
--- a/engine/rig/src/dmsdk/rig/rig.h
+++ b/engine/rig/src/dmsdk/rig/rig.h
@@ -108,11 +108,11 @@ namespace dmRig
     struct RigModelVertex
     {
         float pos[3];
-        float normal[3];
-        float tangent[4];
-        float color[4];
-        float uv0[2];
-        float uv1[2];
+        int8_t normal[4];
+        // int16_t tangent[4];
+        // int16_t color[4];
+        int16_t uv0[2];
+        // int16_t uv1[2];
     };
 
     // Can we not use the skeleton directly?
diff --git a/engine/sound/src/devices/device_wasapi.cpp b/engine/sound/src/devices/device_wasapi.cpp
index 3253f93ef2..a1253cadea 100644
--- a/engine/sound/src/devices/device_wasapi.cpp
+++ b/engine/sound/src/devices/device_wasapi.cpp
@@ -386,6 +386,7 @@ namespace dmDeviceWasapi
 
         info->m_MixRate = device->m_MixFormat->nSamplesPerSec;
         info->m_FrameCount = device->m_FrameCount;
+        dmLogInfo("<!> DeviceWasapi: info->m_MixRate: %d, info->m_FrameCount: %d", info->m_MixRate, info->m_FrameCount);
     }
 
     static void DeviceWasapiStart(dmSound::HDevice _device)
diff --git a/engine/tools/src/gdc/main.cpp b/engine/tools/src/gdc/main.cpp
index d5d28454fe..42419d96ec 100644
--- a/engine/tools/src/gdc/main.cpp
+++ b/engine/tools/src/gdc/main.cpp
@@ -106,7 +106,11 @@ int main(int argc, char *argv[])
     window_params.m_Title = "gdc";
     window_params.m_PrintDeviceInfo = false;
     window_params.m_OpenGLVersionHint = 33;
-    window_params.m_GraphicsApi = dmPlatform::PLATFORM_GRAPHICS_API_OPENGL;
+    if (dmGraphics::GetInstalledAdapterFamily() == dmGraphics::ADAPTER_FAMILY_OPENGLES) {
+        window_params.m_GraphicsApi = dmPlatform::PLATFORM_GRAPHICS_API_OPENGLES;
+    } else {
+        window_params.m_GraphicsApi = dmPlatform::PLATFORM_GRAPHICS_API_OPENGL;
+    }
     window_params.m_ContextAlphabits = 8;
 
     dmPlatform::HWindow window = dmPlatform::NewWindow();
diff --git a/engine/tools/src/gdc/wscript b/engine/tools/src/gdc/wscript
index 85d3dbb728..6226e31e13 100644
--- a/engine/tools/src/gdc/wscript
+++ b/engine/tools/src/gdc/wscript
@@ -8,6 +8,9 @@ def configure(conf):
 def build(bld):
     graphics_lib = 'GRAPHICS OPENGL'
     graphics_adapter = 'GraphicsAdapterOpenGL'
+    if bld.env.PLATFORM in ('arm64-linux'):
+        graphics_lib = 'GRAPHICS_OPENGLES OPENGLES'
+        graphics_adapter = 'GraphicsAdapterOpenGLES'
     if waflib.Options.options.with_vulkan:
         graphics_lib = 'GRAPHICS_VULKAN'
         graphics_adapter = 'GraphicsAdapterVulkan'
diff --git a/engine/tools/wscript b/engine/tools/wscript
index 960fb697a9..dbca94d4e3 100755
--- a/engine/tools/wscript
+++ b/engine/tools/wscript
@@ -45,8 +45,11 @@ def configure(conf):
         conf.env.append_value('LINKFLAGS', ['-framework', 'Cocoa', '-framework', 'OpenGL', '-framework', 'AGL', '-framework', 'IOKit', '-framework', 'Carbon', '-framework', 'CoreVideo', '-framework', 'QuartzCore'])
     elif platform == 'x86_64-ios':
         pass
-    elif platform in ["x86_64-linux", "arm64-linux"]:
+    elif platform == 'x86_64-linux':
         conf.env['LIB_X'] = ['Xext', 'X11', 'Xi', 'GL', 'GLU']
+    elif platform == 'arm64-linux':
+        print("<!> graphics/tools: GLES LINKFLAGS")
+        conf.env['LIB_X'] = ['Xext', 'X11', 'Xi', 'EGL', 'GLESv1_CM', 'GLESv2']
     elif "win32" in platform:
         conf.env.append_value('LINKFLAGS', ['opengl32.lib', 'user32.lib', 'shell32.lib', 'xinput9_1_0.lib', 'dinput8.lib', 'dxguid.lib'])
     elif re.match('arm.*?android', platform):
diff --git a/patch-and-copy-angle b/patch-and-copy-angle
new file mode 100644
index 0000000000..836470a292
--- /dev/null
+++ b/patch-and-copy-angle
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+git diff --cached --binary > angle.patch && cp angle.patch /mnt/c/Users/Artsiom/Documents/Defold/ANGLE/ci-test/all.patch
+cp packages/glfw-3.4-arm64-linux-GLES.tar.gz /mnt/c/Users/Artsiom/Documents/Defold/ANGLE/ci-test/glfw-3.4-arm64-linux.tar.gz && rm -f angle.patch
+# cp packages/astcenc-8b0aa01-arm64-linux-GLES.tar.gz /mnt/c/Users/Artsiom/Documents/Defold/ANGLE/ci-test/
diff --git a/scripts/build.py b/scripts/build.py
index efd07c4d5f..b29298c257 100755
--- a/scripts/build.py
+++ b/scripts/build.py
@@ -1010,6 +1010,9 @@ class Configuration(object):
         if self.target_platform in ('x86_64-macos','arm64-macos','arm64-ios','x86_64-ios') and 'linux' == sys.platform:
             strip = os.path.join(sdkfolder, 'linux', sdk.PACKAGES_LINUX_CLANG, 'bin', 'x86_64-apple-darwin19-strip')
 
+        if self.target_platform == 'arm64-linux' and self.host != self.target_platform:
+            strip = '/usr/bin/aarch64-linux-gnu-strip'
+
         run.shell_command("%s %s" % (strip, path))
         return True
 
@@ -1107,6 +1110,7 @@ class Configuration(object):
         supported_tests['x86_64-win32'] = ['win32', 'x86_64-win32', 'arm64-nx64', 'x86_64-ps4', 'x86_64-ps5']
         supported_tests['arm64-macos'] = ['x86_64-macos', 'arm64-macos', 'wasm-web', 'js-web']
         supported_tests['x86_64-macos'] = ['x86_64-macos', 'wasm-web', 'js-web']
+        supported_tests['x86_64-linux'] = ['arm64-linux']
 
         return self.target_platform in supported_tests.get(self.host, []) or self.host == self.target_platform
 
@@ -1369,9 +1373,12 @@ class Configuration(object):
         test_dir = join(self.defold_root, 'com.dynamo.cr/com.dynamo.cr.bob.test')
 
         sha1 = self._git_sha1()
+        print("<!> path is %s" % os.path.join(self.dynamo_home, 'archive', sha1))
         if os.path.exists(os.path.join(self.dynamo_home, 'archive', sha1)):
+            print("<!> copy.sh")
             run.env_shell_command(self._form_env(), "./scripts/copy.sh", cwd=bob_dir)
         else:
+            print("<!> copy_local_bob_artefacts")
             self.copy_local_bob_artefacts()
 
         env = self._form_env()
@@ -2137,8 +2144,17 @@ class Configuration(object):
 
 
     def upload_to_archive(self, src_file, dst_path):
-        url = join(self.get_archive_path(), dst_path).replace("\\", "/")
-        self.upload_to_s3(src_file, url)
+        # url = join(self.get_archive_path(), dst_path).replace("\\", "/")
+        # self.upload_to_s3(src_file, url)
+        print("<!> upload_to_archive: src_file %s, dst_path %s" % (src_file, dst_path))
+        dst = os.path.join(self.defold_root, self.dynamo_home, "archive")
+        print("                       dst %s" % dst)
+        self._mkdirs(os.path.join(dst, os.path.dirname(dst_path)))
+        print("                       mkdirs %s" % os.path.join(dst, os.path.dirname(dst_path)))
+        if os.path.isdir(src_file):
+            self._copy_tree(src_file, os.path.join(dst, dst_path))
+        else:
+            self._copy(src_file, os.path.join(dst, dst_path))
 
 
     def download_from_s3(self, path, url):
diff --git a/share/ext/common.sh b/share/ext/common.sh
index a67c99afd4..40f1ed5a8f 100755
--- a/share/ext/common.sh
+++ b/share/ext/common.sh
@@ -386,10 +386,11 @@ function cmi_setup_cc() {
             export CXXFLAGS="${CXXFLAGS} --target=aarch64-unknown-linux-gnu -fPIC"
             export CPPFLAGS="${CPPFLAGS} --target=aarch64-unknown-linux-gnu -fPIC"
 
-            export CC=$(which clang)
-            export CXX=$(which clang++)
-            export AR=$(which ar)
-            export RANLIB=$(which ranlib)
+            echo "A"
+            export CC=/home/linux/defold-engine/tmp/dynamo_home/ext/SDKs/x86_64-linux/clang-16.0.0/bin/clang
+            export CXX=/home/linux/defold-engine/tmp/dynamo_home/ext/SDKs/x86_64-linux/clang-16.0.0/bin/clang++
+            export AR=/home/linux/defold-engine/tmp/dynamo_home/ext/SDKs/x86_64-linux/clang-16.0.0/bin/ar
+            export RANLIB=/home/linux/defold-engine/tmp/dynamo_home/ext/SDKs/x86_64-linux/clang-16.0.0/bin/ranlib
             export CPP="${CC} -E"
             ;;
 
diff --git a/share/ext/glfw/arm64-linux.cmake b/share/ext/glfw/arm64-linux.cmake
new file mode 100644
index 0000000000..e782584d29
--- /dev/null
+++ b/share/ext/glfw/arm64-linux.cmake
@@ -0,0 +1,7 @@
+set(CMAKE_SYSTEM_NAME    Linux)
+set(CMAKE_SYSTEM_PROCESSOR aarch64)
+set(CMAKE_C_COMPILER     "aarch64-linux-gnu-gcc")
+set(CMAKE_CXX_COMPILER   "aarch64-linux-gnu-g++")
+set(CMAKE_FIND_ROOT_PATH /usr/aarch64-linux-gnu)
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
+set(CPACK_DEBIAN_PACKAGE_ARCHITECTURE arm64)
diff --git a/share/ext/glfw/build_glfw.sh b/share/ext/glfw/build_glfw.sh
index e019f0634b..76fed11ce8 100755
--- a/share/ext/glfw/build_glfw.sh
+++ b/share/ext/glfw/build_glfw.sh
@@ -69,6 +69,10 @@ case $PLATFORM in
     x86_64-macos)
         CMAKE_FLAGS="-DCMAKE_OSX_ARCHITECTURES=x86_64 -DCMAKE_OSX_DEPLOYMENT_TARGET=${OSX_MIN_SDK_VERSION} ${CMAKE_FLAGS}"
         ;;
+    arm64-linux)
+        # Need to setup clang on linux
+        cmi_setup_cc $PLATFORM
+        ;;
 esac
 
 function cmi_unpack() {
diff --git a/share/extender/build_input.yml b/share/extender/build_input.yml
index 781250fdee..7e49752b60 100644
--- a/share/extender/build_input.yml
+++ b/share/extender/build_input.yml
@@ -11,7 +11,7 @@ context:
                   "engine_service","engine_service_null","libengine_service.lib","libengine_service_null.lib",
                   "script_box2d", "script_box2d_null", "libscript_box2d.lib", "libscript_box2d_null.lib"]
     allowedSymbols: ["DefaultSoundDevice", "NullSoundDevice", "AudioDecoderWav", "CrashExt", "AudioDecoderStbVorbis", "GraphicsAdapterWebGPU",
-                    "ProfilerExt", "GraphicsAdapterOpenGL", "GraphicsAdapterVulkan", "GraphicsAdapterNull",
+                    "ProfilerExt", "GraphicsAdapterOpenGL", "GraphicsAdapterOpenGLES", "GraphicsAdapterVulkan", "GraphicsAdapterNull",
                     "LiveUpdateExt", "ScriptBox2DExt", "ScriptImageExt", "ScriptModelExt", "ScriptTypesExt"]
     defines: ['DLIB_LOG_DOMAIN="{{extension_name_upper}}"']
 
